src/java.base/share/classes/java/util/zip/ZipOutputStream.java

```diff
@@ -182,6 +210,60 @@ public void setLevel(int level) {
      * The default compression method will be used if no compression method
      * was specified for the entry, and the current time will be used if
      * the entry has no set modification time.
+     *
+     * @apiNote
+     * In general it is not safe to directly write a {@code ZipEntry} obtained
+     * from {@link ZipInputStream#getNextEntry()},  {@link ZipFile#entries()},
+     * {@link ZipFile#getEntry(String)} or {@link ZipFile#stream()} with this
+     * method to a {@code ZipOutputStream} as follows:
+     *
+     * <blockquote><pre>
+     * ZipEntry entry;
+     * ZipInputStream zis = new ZipInputStream(...);
+     * ZipOutputStream zos = new ZipOutputStream(...);
+     * while((entry = zis.getNextEntry()) != null) {
+     *     zos.putNextEntry(entry);
+     *     zis.transferTo(zos);
+     * }
+     * </pre></blockquote>
+     *
+     * The problem with this code is that the zip file format does not record
+     * the compression level used for deflation in its entries. In general, it
+     * doesn't even mandate a predefined compression ratio per compression
+     * level. Therefore the compressed size recorded in a {@code ZipEntry}
+     * read from a zip file might differ from the new compressed size produced
+     * by the receiving {@code ZipOutputStream}. Such a difference will result
+     * in a {@link ZipException} with the following message:
+     *
+     * <blockquote><pre>
+     * java.util.zip.ZipException: invalid entry compressed size (expected 12 but got 7 bytes)
+     * </pre></blockquote>
+     *
+     * The correct way of copying all entries from one zip file into another
+     * requires the creation of a new {@code ZipEntry} or at least resetting of the
+     * compressed size field. E.g.:
+     *
+     * <blockquote><pre>
+     * while((entry = zis.getNextEntry()) != null) {
+     *     entry.setCompressedSize(-1);
+     *     zos.putNextEntry(entry);
+     *     zis.transferTo(zos);
+     * }
+     * </pre></blockquote>
+     *
+     * @implNote
+     * This implementation supports the system property
+     * {@code jdk.util.zip.ZipEntry.compressedSizeHandling} which can be used
```

Comment:
Might be good to use the [`{@systemProperty ...}` inline tag](https://docs.oracle.com/en/java/javase/15/docs/specs/javadoc/doc-comment-spec.html#systemproperty) added in JDK 12.